jormungandr/src/fragment/expirations.rs

```diff
@@ -0,0 +1,104 @@
+use slab::Slab;
+use std::{
+    collections::{BTreeMap, HashSet},
+    time::{Duration, Instant},
+};
+
+/// A synchronous alternative to tokio's `DelayQueue` intended to be
+/// periodically polled for expired entries.
+pub struct Expirations<T> {
+    slab: Slab<ExpirationEntry<T>>,
+    sorted: BTreeMap<SlotId, HashSet<Key>>,
```

Comment:
The B-tree not ideal for popping entries, which is likely the most frequent operation. But since this queue must also support removals and reschedulings, a simple binary heap would be a poor fit too.